File: java/com/joshmanisdabomb/lcc/fluid/OilFluid.java
package com.joshmanisdabomb.lcc.fluid;

import net.minecraft.block.BlockState;
import net.minecraft.block.Blocks;
import net.minecraft.block.FireBlock;
import net.minecraft.block.FlowingFluidBlock;
import net.minecraft.fluid.Fluid;
import net.minecraft.fluid.IFluidState;
import net.minecraft.state.StateContainer;
import net.minecraft.tags.FluidTags;
import net.minecraft.util.Direction;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.IBlockReader;
import net.minecraft.world.IWorld;
import net.minecraft.world.IWorldReader;
import net.minecraftforge.fluids.ForgeFlowingFluid;

public class OilFluid extends ForgeFlowingFluid {

    private final boolean source;

    public OilFluid(boolean source, Properties properties) {
        super(properties);
        this.source = source;
    }

    @Override
    public boolean isSource(IFluidState state) {
        return this.source;
    }

    @Override
    public int getLevel(IFluidState state) {
        return this.source ? 8 : state.get(LEVEL_1_8);
    }

    @Override
    protected void fillStateContainer(StateContainer.Builder<Fluid, IFluidState> builder) {
        super.fillStateContainer(builder);
        if (!this.source) builder.add(LEVEL_1_8);
    }

    @Override
    public int getTickRate(IWorldReader world) {return 20;}

    @Override
    protected void flowInto(IWorld world, BlockPos pos, BlockState blockState, Direction d, IFluidState fluidState) {
        IFluidState fluid = world.getFluidState(pos);
        if (blockState.getBlock() == Blocks.FIRE || blockState.getBlock() instanceof FlowingFluidBlock) {
            if (blockState.getBlock() == Blocks.FIRE || fluid.isTagged(FluidTags.LAVA)) {
                world.setBlockState(pos.offset(d.getOpposite()), net.minecraftforge.event.ForgeEventFactory.fireFluidPlaceBlockEvent(world, pos.offset(d.getOpposite()), pos.offset(d.getOpposite()), Blocks.FIRE.getDefaultState().with(FireBlock.AGE, 15)), 3);
            }
            return;
        }
        super.flowInto(world, pos, blockState, d, fluidState);
    }

    @Override
    protected boolean func_215665_a(IFluidState state, IBlockReader world, BlockPos pos, Fluid fluidIn, Direction direction) {
        return false;
    }

}

File: lcc-content/src/main/kotlin/com/joshmanisdabomb/lcc/fluid/OilFluid.kt
package com.joshmanisdabomb.lcc.fluid

import com.joshmanisdabomb.lcc.directory.LCCBlocks
import com.joshmanisdabomb.lcc.directory.LCCFluids
import com.joshmanisdabomb.lcc.directory.LCCItems
import com.joshmanisdabomb.lcc.trait.LCCFluidTrait
import net.minecraft.block.*
import net.minecraft.fluid.FlowableFluid
import net.minecraft.fluid.Fluid
import net.minecraft.fluid.FluidState
import net.minecraft.state.StateManager
import net.minecraft.tag.BlockTags
import net.minecraft.tag.FluidTags
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.Direction
import net.minecraft.world.BlockView
import net.minecraft.world.WorldAccess
import net.minecraft.world.WorldView

class OilFluid(val source: Boolean) : FlowableFluid(), LCCFluidTrait {

    val colors = FloatArray(3) { 0.03f }

    override fun getStill() = LCCFluids.oil_still

    override fun getFlowing() = LCCFluids.oil_flowing

    override fun matchesType(fluid: Fluid) = fluid == still || fluid == flowing

    override fun getBucketItem() = LCCItems.oil_bucket

    override fun isStill(state: FluidState) = source

    override fun getLevel(state: FluidState) = if (source) 8 else state[LEVEL]

    override fun appendProperties(builder: StateManager.Builder<Fluid, FluidState>) = super.appendProperties(builder).also { if (!source) builder.add(LEVEL) }

    override fun getFlowSpeed(world: WorldView) = 1

    override fun getLevelDecreasePerBlock(world: WorldView) = 7

    override fun canBeReplacedWith(state: FluidState, world: BlockView, pos: BlockPos, fluid: Fluid, direction: Direction) = false

    override fun getTickRate(world: WorldView) = 25

    override fun getBlastResistance() = 2f

    override fun toBlockState(state: FluidState) = LCCBlocks.oil.defaultState.with(FluidBlock.LEVEL, getBlockStateLevel(state));

    override fun isInfinite() = false

    override fun beforeBreakingBlock(world: WorldAccess, pos: BlockPos, state: BlockState) {
        Block.dropStacks(state, world, pos, if (state.hasBlockEntity()) world.getBlockEntity(pos) else null)
    }

    override fun flow(world: WorldAccess, pos: BlockPos, state: BlockState, direction: Direction, fluidState: FluidState) {
        when {
            state.isIn(BlockTags.FIRE) || state.fluidState.isIn(FluidTags.LAVA) -> {
                val fire = AbstractFireBlock.getState(world, pos.offset(direction.opposite)).run { if (this == Blocks.FIRE) with(FireBlock.AGE, 15) else this }
                world.setBlockState(pos.offset(direction.opposite), fire, 3)
                world.blockTickScheduler.schedule(pos, fire.block, world.random.nextInt(5) + 2)
            }
            state.block is FluidBlock -> return
            else -> return super.flow(world, pos, state, direction, fluidState)
        }
    }

    override fun lcc_fogColor() = colors

    override fun lcc_fogDensity() = 0.7f//RenderSystem.fogMode(GlStateManager.FogMode.EXP2).let { 0.7f }

}
